adi/rx_tx.py

```diff
@@ -255,84 +252,62 @@ def __rx_unbuffered_data(self):
 
         return x
 
-    def __rx_buffered_data(self):
+    def __rx_buffered_data(self) -> Union[List[np.ndarray], np.ndarray]:
+        """__rx_buffered_data: Read data from RX buffer
+
+        Returns:
+            List of numpy arrays containing the data from the RX buffer that are
+            channel interleaved
+        """
         if not self.__rxbuf:
             self._rx_init_channels()
         self.__rxbuf.refill()
-        data = []
 
-        iio_enabled_channels = []
-        if self._complex_data is True:
+        data_channel_interleaved = []
+        ecn = []
+        if self._complex_data:
             for m in self.rx_enabled_channels:
-                iio_enabled_channels.append(self._rxadc.find_channel(self._rx_channel_names[m * 2]))
-                iio_enabled_channels.append(self._rxadc.find_channel(self._rx_channel_names[(m * 2) + 1]))
+                ecn.extend(
+                    (self._rx_channel_names[m * 2], self._rx_channel_names[m * 2 + 1])
+                )
         else:
-            for m in self.rx_enabled_channels:
-                iio_enabled_channels.append(self._rxadc.find_channel(self._rx_channel_names[m]))
-
-        iio_enabled_channels.sort(key=lambda ch: ch.index) # sort channels by index
-
-        data = self.__rxbuf.read() # read buffer from iio
-        # initialize and build dtype tuple for np.frombuffer
-        npdtype = ()
-        for ec in iio_enabled_channels:
-            df = ec.data_format
-            fmt = ">" if df.is_be is True else "<"
-            fmt += "i" if df.is_signed is True else "u"
-            fmt += str(df.length // 8)  # create format string -
-            npdtype = (*npdtype, fmt)
-        x = np.frombuffer(data, dtype=npdtype)
-
-        stride = len(iio_enabled_channels)
-        for idx, ec in enumerate(iio_enabled_channels):
-            df = ec.data_format
-            x[idx::stride] = np.right_shift(x[idx::stride], df.shift)
-            x[idx::stride] = np.bitwise_and(x[idx::stride], (1 << df.length) - 1)
-
-        if not self._rx_stack_interleaved and stride>1:
-            stacked_data = []
-            for idx, ec in enumerate(iio_enabled_channels):
-                stacked_data.extend(x[idx::stride])
-            x = stacked_data
-        return x
+            ecn = [self._rx_channel_names[m] for m in self.rx_enabled_channels]
+
```

Comment:
enable channels need to be sorted by index ... if in user application you enable channels such as:
[ch2,ch0,ch1]
but the samples come interleaved as 
ch0,ch1,ch2,ch0,ch1.ch2

You will apply ch2 format to ch0 and so on. This is important for multitype.
This is implemented here: 
https://github.com/analogdevicesinc/pyadi-iio/blob/rx_tx-dynamic-channel-type-conversion/adi/rx_tx.py#L273
However my implementation makes use of iio_channels instead of iio channel names which makes things a little "dirtier"

Reply:
EDIT! Actually, nevermind ! If you read per channel first, you don't care about all this ! Just read channel using libiio API and apply it's own formatting. No changes needed here !